src/cargo/ops/cargo_rustc/context.rs

```diff
@@ -360,19 +357,30 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
         match self.target_metadata(unit) {
             Some(ref metadata) => format!("{}{}", unit.target.crate_name(),
                                           metadata.extra_filename),
-            None if unit.target.allows_underscores() => {
-                unit.target.name().to_string()
-            }
-            None => unit.target.crate_name(),
+            None => self.bin_stem(unit),
+        }
+    }
+
+    fn bin_stem(&self, unit: &Unit) -> String {
+        let bin = if unit.target.allows_underscores() {
+            unit.target.name().to_string()
+        } else {
+            unit.target.crate_name()
+        };
+        if unit.profile.test {
+            bin + "-test"
```

Comment:
Yeah. This was needed to separate test binaries from true binaries. Previously they were separated by metadata hash. The move-targets-up code was linking them on top of each other in this code. One option was to replicate the -{metadata} behavior, but I thought it simpler to just do "-test" instead now that there was a clear way of separating them.